pkg/splunk/enterprise/afwscheduler.go

```diff
@@ -981,6 +981,143 @@ func afwGetReleventStatefulsetByKind(cr splcommon.MetaObject, client splcommon.C
 	return sts
 }
 
+// getPodExecClient returns the client object used to execute pod exec commands
+func getPodExecClient(client splcommon.ControllerClient, cr splcommon.MetaObject, cmdStr string, targetPodName string) *PodExecClient {
+	scopedLog := log.WithName("getPodExecClient").WithValues("crName", cr.GetName(), "namespace", cr.GetNamespace())
+
+	// get the admin password from the namespace scoped secret
+	adminPwd, err := getAdminPasswordFromSecret(client, cr)
+	if err != nil {
+		scopedLog.Error(err, "unable to get admin password")
+		return nil
+	}
+
+	return &PodExecClient{
+		client:        client,
+		cr:            cr,
+		targetPodName: targetPodName,
+		adminPwd:      adminPwd,
+		command:       fmt.Sprintf(cmdStr, adminPwd),
+	}
+}
+
+// getIdxcPlayBookContext returns the idxc playbook context
+func getIdxcPlayBookContext(client splcommon.ControllerClient, cr splcommon.MetaObject, afwPipeline *AppInstallPipeline, podName string, podExecClient PodExecClientImpl) *IdxcPlayBookContext {
+	return &IdxcPlayBookContext{
+		client:        client,
+		cr:            cr,
+		afwPipeline:   afwPipeline,
+		targetPodName: podName,
+		podExecClient: podExecClient,
+	}
+}
+
+// getSHCPlayBookContext returns the shc playbook context
+func getSHCPlayBookContext(client splcommon.ControllerClient, cr splcommon.MetaObject, afwPipeline *AppInstallPipeline, podName string) *SHCPlayBookContext {
+	return &SHCPlayBookContext{
+		client:        client,
+		cr:            cr,
+		afwPipeline:   afwPipeline,
+		targetPodName: podName,
+	}
+}
+
+// getPlayBookContext returns the context for running playbook
+func getPlayBookContext(client splcommon.ControllerClient, cr splcommon.MetaObject, afwPipeline *AppInstallPipeline, podName string, kind string, podExecClient PodExecClientImpl) PlayBookImpl {
+
+	switch kind {
+	case "ClusterMaster":
+		return getIdxcPlayBookContext(client, cr, afwPipeline, podName, podExecClient)
+	// TODO: gaurav - implement SHC playbook
+	//case "SearchHeadCluster":
+	//	return getSHCPlayBookContext(client, cr, afwPipeline, podName)
+	default:
+		return nil
+	}
+}
+
+// runPodExecCommand runs the commands related to idxc bundle push
+func (podExecClient *PodExecClient) runPodExecCommand() (string, string, error) {
+	if podExecClient == nil {
+		err := fmt.Errorf("podExecClient is nil")
+		return "", "", err
+	}
+
+	streamOptions := &remotecommand.StreamOptions{
+		Stdin: strings.NewReader(podExecClient.command),
```

Comment:
Can we make the command as input to this method, that way, the user has the option to re-use the same client with multiple commands. One of my earlier comments was to pass the bundle push command as part of the `isBundlePushComplete`. So, the caller fo the APIs need not worry about setting the command.

Reply:
I get your point but I think for the exact reason that users can re-use this API with different command/s, I embedded the command in the client itself, they can just change the command in the client and can re-use it. In either case we are relying on the caller of this API to set the command to be called in the podExec command.
Also, about your second point of passing the command in the `isBundlePushComplete` API, as I mentioned before too, I could not pass it inside the API since we require admin password to create the whole command string. Now, we need this sort of conversion i.e. `podExecClient := idxcPlayBookContext.podExecClient.(*PodExecClient)` to convert the generic `PodExecClientImpl` to `PodExecClient` so that we can access the `adminPwd` field. And if you are trying to use `mockPodExecClient` in UT, it will crash at the point of conversion to `PodExecClient`. Hence to cover the lines in `isBundlePushComplete`, I had to set the command before calling the podExec API. But let's discuss it more offline and then maybe we can reach a common ground :)

Reply:
NVM, I edited my previous comment, I got what you are saying and will change the code accordingly!